river/command/map.zig

```diff
@@ -31,6 +31,21 @@ const SwitchMapping = @import("../SwitchMapping.zig");
 const Switch = @import("../Switch.zig");
 const Seat = @import("../Seat.zig");
 
+const KeycodeMatch = struct {
+    keysym: xkb.Keysym,
+    keycode: xkb.Keycode,
+};
+
+/// Iterate through keycodes in the keymap to find ones matching the specified keysym.
+/// If more than one keycode matches, use the smallest.
```

Comment:
It'd be nice to explain why the smallest is used in this comment, perhaps with an example.

Reply:
Sure I'll add an example to the comment.

What do you think of making keycodes a linked list and add all found keycodes to the mapping? This would probably be the most "correct" way.

The cases where I found second keycodes were all quite high keycodes that do not appear on normal keyboards, e.g. `KEY_KPLEFTPAREN` or even more obscure ones. (This was with French AZERTY layout). But users could have custom layouts where they have a keysym on multiple keys and they want the mapping for all of them.

Reply:
Hmm, I'm not a huge fan of introducing more complexity for an edge case I'm not sure we'll ever see in practice.

I was just looking around the xkbcommon API a bit and found the `xkb_keymap_key_by_name()` function which translates a string name to a keycode. Its implementation looks similar to what you have here, but with something additional about aliases. Perhaps we should just use that function?
```c
XKB_EXPORT xkb_keycode_t
xkb_keymap_key_by_name(struct xkb_keymap *keymap, const char *name)
{
    struct xkb_key *key;
    xkb_atom_t atom;

    atom = xkb_atom_lookup(keymap->ctx, name);
    if (atom) {
        xkb_atom_t ratom = XkbResolveKeyAlias(keymap, atom);
        if (ratom)
            atom = ratom;
    }
    if (!atom)
        return XKB_KEYCODE_INVALID;

    xkb_keys_foreach(key, keymap) {
        if (key->name == atom)
            return key->keycode;
    }

    return XKB_KEYCODE_INVALID;
}
```

Reply:
It seems to me that the simplest way forward with this would be to put everything in libxkbcommon's hands and use `xkb_keymap_key_by_name()` for keycode based mappings and `xkb_keysym_from_name()` for keysym based mappings. I don't think we should ever parse or store a keysym for keycode based mappings as doing a keysym to keycode is potentially not one-to-one and seems overcomplex to me.

Reply:
Hmm, that might be a bad idea actually. It's not clear to me that keycodes consistently have good user-facing names, it seems to depend on the keymap.

Reply:
Hm then we could just do a direct keycode mapping like sway's `bindcode`?

The huge advantage of keysym-to-keycode mapping IMO is that the user doesn't need to look up keycodes but just add the flag to an existing mapping. I think this was the reason why sway decided to add `--to-code` although it was not truly necessary because they already had `bindcode`.

Reply:
Did a bit more poking around. It looks like since the only input backend we/wlroots currently support is libinput which uses evdev the keycodes we receive from wlroots are always linux input event codes as defined in `linux/input-event-codes.h`. We already use these codes for mouse buttons, so I wouldn't mind using them here as well.

Thanks for reminding me that sway has bincode as well, I haven't used sway in quite a while. It looks like the way they implement it isn't very user-friendly though as it takes an xkb keycode number directly. We could instead make our version accept linux input event code names as I mentioned above, parsing with `libevdev_event_code_from_name()` and adding 8 to convert from linux input event code to xkb keycode.

As for whether we should have an equivalent to sway's `bindsym --to-code`, `bindcode`, or both let's remember that converting keysym to keycode as is done by `bindsym --to-code` can be ambiguous. This makes me think that we must eventually implement a `bindcode` equivalent in order to support the maximal number of use-cases. Furthermore, if we implement a `bindcode` equivalent that is not painful to use, I don't think we need a redundant `bindsym --to-code` equivalent.

What are your thoughts there? Thanks for helping me figure this stuff out :)

Reply:
Yeah, I thought about the same when implementing this. I was also sceptical about the redundancy of having both methods. Eventually, I chose `-to-code` over a direct keycode mapping because of the simplicity for the user, despite the potential ambiguity of the translation.

I do agree that evdev code names are nicer than the plain numbers sway uses for `bindcode`. But still users would need `sudo evtest` or something similar to find the event code name in a lot of cases. 

True, to support every use case, we need an equivalent to sway's `bindcode`. But do these use cases ever emerge? I'm not sure...

My concern is that with exclusive direct keycode mappings (be it with event code names or xkb keycodes), we end up with lots of complaints from people who doesn't want to lookup the event code names for all mappings. Like here https://github.com/swaywm/sway/issues/2999 and here https://github.com/swaywm/sway/pull/3058 .

Me personally, I'm fine with either solution (or even both) and I don't mind implementing a direct keycode mapping if you decide so.

> What are your thoughts there? Thanks for helping me figure this stuff out :)

You're welcome, I'm happy to be a part of this :)

Reply:
Cool, I think I'd like to stick to binding event codes directly by name and see how that works out. The implementation is far simpler and there's less that can go wrong from a technical perspective. It seems that emersion's initial reaction matched mine as well.

The core complaint in those sway discussions seems to be the readability of the config file, which I think will be mostly alleviated by using `KEY_A` instead of `30` for example. We could probably drop the `KEY_` prefix as well, not 100% confident that won't cause issues though.

As for the `riverctl` interface, I'm thinking it'd be more user friendly to have a separate `map-keycode` command, which makes me think we should rename `map` to `map-keysym`.
```
       map-keysym [-release|-repeat] mode modifiers keysym command
       map-keycode [-release|-repeat] mode modifiers keycode command
```

If this proves to be too hard to use, I'd rather explore adding a way for riverctl to print out the keysyms/keycodes being pressed before adding (flawed) keysym to keycode translation.

Reply:
I think this would be very usable. If we dropped the `KEY_` prefix and make it case-insensitive, it would result in the same name as the keysym for most mappings in the default init file. So users with other layouts could use the default init file without changing much.

But then we maybe should do the same with mouse buttons for consistency.

Reply:
> I think this would be very usable. If we dropped the `KEY_` prefix and make it case-insensitive, it would result in the same name as the keysym for most mappings in the default init file. So users with other layouts could use the default init file without changing much.

Sounds good to me, using `std.ascii` to uppercase whatever string the user provides and then prepending `KEY_` should be robust. Good point about the mouse buttons, let's leave that for a follow up patch if this one goes well.

The main challenge remaining is documentation and spending some more thought on whether `map-keysym`/`map-keycode` are the best names we can come up with. `map-keycode` *kinda* sounds like it expects a numerical code but I don't have a better suggestion.

Reply:
> Sounds good to me, using `std.ascii` to uppercase whatever string the user provides and then prepending `KEY_` should be robust. Good point about the mouse buttons, let's leave that for a follow up patch if this one goes well.

Ah yeah thinking about this, I remember there was a feature request some time ago (I think on IRC) to map mouse events to commands. With a dedicated `map-keycode` and corresponding `KeycodeMapping` type, I think this could become feasible. If we want to leave the door open for this we cannot drop the `BTN_` *and* `KEY_` prefixes. 

> The main challenge remaining is documentation and spending some more thought on whether `map-keysym`/`map-keycode` are the best names we can come up with. `map-keycode` _kinda_ sounds like it expects a numerical code but I don't have a better suggestion.

The only alternative name I can come up with on the fly is `map-event` or. Or we simplify as `map-sym` and `map-code`. I also feel like we could leave `map` as is.

Reply:
or even `map-evdev`

Reply:
I think leaving `map` as is and introducing `map-code` is what I'd like to go with. There's no strong need for a breaking change here and `map-code` avoids the term "keycode" which usually refers to an xkbcommon keycode not a linux input event code.

Reply:
Yes I agree!